Inline stroking a single PathSegment

shallawa opened the pull request:
#### 3e6f143c6278c7e306454785343cba29355653ad
<pre>
Inline stroking a single PathSegment
<a href="https://bugs.webkit.org/show_bug.cgi?id=262822">https://bugs.webkit.org/show_bug.cgi?id=262822</a>
rdar://116264934

Reviewed by Kimmo Kinnunen.

Like what we do for the single-PathDataLine path, other single-segment path cases
should be handled. Instead of sending a separate SetState message in addition to
the StrokePath to GPUP, we can piggyback the stroke width and color to StrokePath
or StrokePathSegment messages.

* Source/WTF/wtf/PlatformHave.h:
* Source/WTF/wtf/PlatformUse.h:
* Source/WebCore/platform/graphics/BifurcatedGraphicsContext.cpp:
(WebCore::BifurcatedGraphicsContext::fillPathSegment):
(WebCore::BifurcatedGraphicsContext::strokePathSegment):
* Source/WebCore/platform/graphics/BifurcatedGraphicsContext.h:
* Source/WebCore/platform/graphics/GraphicsContext.h:
* Source/WebCore/platform/graphics/NullGraphicsContext.h:
* Source/WebCore/platform/graphics/Path.cpp:
(WebCore::Path::singleDataLine const): Deleted.
(WebCore::Path::singleArc const): Deleted.
(WebCore::Path::singleQuadCurve const): Deleted.
(WebCore::Path::singleBezierCurve const): Deleted.
(WebCore::Path::definitelySingleLine const): Deleted.
* Source/WebCore/platform/graphics/Path.h:
(WebCore::Path::getSingle const):
(WebCore::Path::isSingleLine const):
(WebCore::Path::singleSegmentIfExists const): Deleted.
* Source/WebCore/platform/graphics/PathImpl.h:
(WebCore::PathImpl::singleSegment const):
(WebCore::PathImpl::singleDataLine const): Deleted.
(WebCore::PathImpl::singleArc const): Deleted.
(WebCore::PathImpl::singleQuadCurve const): Deleted.
(WebCore::PathImpl::singleBezierCurve const): Deleted.
* Source/WebCore/platform/graphics/PathSegment.cpp:
(WebCore::PathSegment::fastBoundingRect const):
(WebCore::PathSegment::createPlatformPath const):
* Source/WebCore/platform/graphics/PathSegment.h:
(WebCore::PathSegment::get const):
* Source/WebCore/platform/graphics/PathStream.cpp:
(WebCore::PathStream::singleDataType const): Deleted.
(WebCore::PathStream::singleDataLine const): Deleted.
(WebCore::PathStream::singleArc const): Deleted.
(WebCore::PathStream::singleQuadCurve const): Deleted.
(WebCore::PathStream::singleBezierCurve const): Deleted.
* Source/WebCore/platform/graphics/PathStream.h:
* Source/WebCore/platform/graphics/cairo/PathCairo.cpp:
(WebCore::PathCairo::createPlatformPath):
* Source/WebCore/platform/graphics/cairo/PathCairo.h:
* Source/WebCore/platform/graphics/cg/GraphicsContextCG.cpp:
(WebCore::GraphicsContextCG::fillPathWithGradient):
(WebCore::GraphicsContextCG::fillPathWithGradientAndShadow):
(WebCore::GraphicsContextCG::fillPath):
(WebCore::GraphicsContextCG::strokePathWithGradient):
(WebCore::GraphicsContextCG::strokePathWithGradientAndShadow):
(WebCore::GraphicsContextCG::strokePath):
(WebCore::GraphicsContextCG::fillPathSegment):
(WebCore::GraphicsContextCG::strokePathSegment):
* Source/WebCore/platform/graphics/cg/GraphicsContextCG.h:
* Source/WebCore/platform/graphics/cg/PathCG.cpp:
(WebCore::PathCG::createPlatformPath):
* Source/WebCore/platform/graphics/cg/PathCG.h:
* Source/WebCore/platform/graphics/displaylists/DisplayListItems.cpp:
(WebCore::DisplayList::StrokeLine::apply const):
* Source/WebCore/platform/graphics/displaylists/DisplayListRecorder.cpp:
(WebCore::DisplayList::Recorder::fillPath):
(WebCore::DisplayList::Recorder::fillPathSegment):
(WebCore::DisplayList::Recorder::strokePath):
(WebCore::DisplayList::Recorder::strokePathSegment):
* Source/WebCore/platform/graphics/displaylists/DisplayListRecorder.h:
* Source/WebCore/platform/graphics/displaylists/DisplayListRecorderImpl.cpp:
(WebCore::DisplayList::RecorderImpl::recordStrokeLineWithColorAndThickness):
(WebCore::DisplayList::RecorderImpl::recordStrokeArcWithColorAndThickness):
(WebCore::DisplayList::RecorderImpl::recordStrokeQuadCurveWithColorAndThickness):
(WebCore::DisplayList::RecorderImpl::recordStrokeBezierCurveWithColorAndThickness):
(WebCore::DisplayList::RecorderImpl::recordStrokePathSegmentWithColorAndThickness):
(WebCore::DisplayList::RecorderImpl::recordStrokePathWithColorAndThickness):
* Source/WebCore/platform/graphics/displaylists/DisplayListRecorderImpl.h:
* Source/WebCore/rendering/svg/RenderSVGPath.cpp:
(WebCore::RenderSVGPath::updateShapeFromElement):
* Source/WebCore/rendering/svg/legacy/LegacyRenderSVGPath.cpp:
(WebCore::LegacyRenderSVGPath::updateShapeFromElement):
* Source/WebKit/GPUProcess/graphics/RemoteDisplayListRecorder.cpp:
(WebKit::RemoteDisplayListRecorder::strokeLineWithColorAndThickness):
(WebKit::RemoteDisplayListRecorder::strokeArcWithColorAndThickness):
(WebKit::RemoteDisplayListRecorder::strokeQuadCurveWithColorAndThickness):
(WebKit::RemoteDisplayListRecorder::strokeBezierCurveWithColorAndThickness):
(WebKit::RemoteDisplayListRecorder::strokePathSegmentWithColorAndThickness):
(WebKit::RemoteDisplayListRecorder::strokePathWithColorAndThickness):
* Source/WebKit/GPUProcess/graphics/RemoteDisplayListRecorder.h:
* Source/WebKit/GPUProcess/graphics/RemoteDisplayListRecorder.messages.in:
* Source/WebKit/WebProcess/GPU/graphics/RemoteDisplayListRecorderProxy.cpp:
(WebKit::RemoteDisplayListRecorderProxy::recordStrokeLineWithColorAndThickness):
(WebKit::RemoteDisplayListRecorderProxy::recordStrokeArcWithColorAndThickness):
(WebKit::RemoteDisplayListRecorderProxy::recordStrokeQuadCurveWithColorAndThickness):
(WebKit::RemoteDisplayListRecorderProxy::recordStrokeBezierCurveWithColorAndThickness):
(WebKit::RemoteDisplayListRecorderProxy::recordStrokePathSegmentWithColorAndThickness):
(WebKit::RemoteDisplayListRecorderProxy::recordStrokePathWithColorAndThickness):
* Source/WebKit/WebProcess/GPU/graphics/RemoteDisplayListRecorderProxy.h:

Canonical link: <a href="https://commits.webkit.org/269504@main">https://commits.webkit.org/269504@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/7d26f756c49e274c4a344fc86b845f87fcd70a39

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/22705 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/379 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/23792 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/24612 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/21020 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/22964 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/1454 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/23235 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/21953 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/22945 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/47/builds/230 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/19742 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/25465 "Built successfully") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/230 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/20578 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/26806 "Passed tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/19772 "Built successfully and passed tests") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/20603 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/20822 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/24646 "Passed tests") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/22071 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/254 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/18104 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/26123 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/197 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/6124 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/5422 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/325 "Built successfully") | | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/24/builds/27405 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/259 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/3/builds/5960 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

kkinnunen-apple commented on Source/WebCore/platform/graphics/cg/GraphicsContextCG.cpp:
```diff
@@ -688,133 +688,187 @@ void GraphicsContextCG::drawPath(const Path& path)
-#if HAVE(CG_CONTEXT_DRAW_PATH_DIRECT)
-    CGContextDrawPathDirect(context, fillRule() == WindRule::EvenOdd ? kCGPathEOFill : kCGPathFill, path.platformPath(), nullptr);
-#else
+    if (!fillGradient()) {
```
nice

kkinnunen-apple commented on Source/WebCore/platform/graphics/GraphicsContext.h:
```diff
@@ -216,6 +216,10 @@ class GraphicsContext {
     virtual void fillPath(const Path&) = 0;
     virtual void strokePath(const Path&) = 0;
     WEBCORE_EXPORT virtual void drawPath(const Path&);
+#if ENABLE(INLINE_PATH_DATA)
```
I hope we could move away from the INLINE_PATH_DATA ifdef, to simplify the interface and correctness testing.

kkinnunen-apple reviewed (approved):
Partly the need to fuse multiple commands together is a failure of the IPC layer. That's one thing to improve..

shallawa committed:
Inline stroking a single PathSegment
https://bugs.webkit.org/show_bug.cgi?id=262822
rdar://116264934

Reviewed by Kimmo Kinnunen.

Like what we do for the single-PathDataLine path, other single-segment path cases
should be handled. Instead of sending a separate SetState message in addition to
the StrokePath to GPUP, we can piggyback the stroke width and color to StrokePath
or StrokePathSegment messages.

* Source/WTF/wtf/PlatformHave.h:
* Source/WTF/wtf/PlatformUse.h:
* Source/WebCore/platform/graphics/BifurcatedGraphicsContext.cpp:
(WebCore::BifurcatedGraphicsContext::fillPathSegment):
(WebCore::BifurcatedGraphicsContext::strokePathSegment):
* Source/WebCore/platform/graphics/BifurcatedGraphicsContext.h:
* Source/WebCore/platform/graphics/GraphicsContext.h:
* Source/WebCore/platform/graphics/NullGraphicsContext.h:
* Source/WebCore/platform/graphics/Path.cpp:
(WebCore::Path::singleDataLine const): Deleted.
(WebCore::Path::singleArc const): Deleted.
(WebCore::Path::singleQuadCurve const): Deleted.
(WebCore::Path::singleBezierCurve const): Deleted.
(WebCore::Path::definitelySingleLine const): Deleted.
* Source/WebCore/platform/graphics/Path.h:
(WebCore::Path::getSingle const):
(WebCore::Path::isSingleLine const):
(WebCore::Path::singleSegmentIfExists const): Deleted.
* Source/WebCore/platform/graphics/PathImpl.h:
(WebCore::PathImpl::singleSegment const):
(WebCore::PathImpl::singleDataLine const): Deleted.
(WebCore::PathImpl::singleArc const): Deleted.
(WebCore::PathImpl::singleQuadCurve const): Deleted.
(WebCore::PathImpl::singleBezierCurve const): Deleted.
* Source/WebCore/platform/graphics/PathSegment.cpp:
(WebCore::PathSegment::fastBoundingRect const):
(WebCore::PathSegment::createPlatformPath const):
* Source/WebCore/platform/graphics/PathSegment.h:
(WebCore::PathSegment::get const):
* Source/WebCore/platform/graphics/PathStream.cpp:
(WebCore::PathStream::singleDataType const): Deleted.
(WebCore::PathStream::singleDataLine const): Deleted.
(WebCore::PathStream::singleArc const): Deleted.
(WebCore::PathStream::singleQuadCurve const): Deleted.
(WebCore::PathStream::singleBezierCurve const): Deleted.
* Source/WebCore/platform/graphics/PathStream.h:
* Source/WebCore/platform/graphics/cairo/PathCairo.cpp:
(WebCore::PathCairo::createPlatformPath):
* Source/WebCore/platform/graphics/cairo/PathCairo.h:
* Source/WebCore/platform/graphics/cg/GraphicsContextCG.cpp:
(WebCore::GraphicsContextCG::fillPathWithGradient):
(WebCore::GraphicsContextCG::fillPathWithGradientAndShadow):
(WebCore::GraphicsContextCG::fillPath):
(WebCore::GraphicsContextCG::strokePathWithGradient):
(WebCore::GraphicsContextCG::strokePathWithGradientAndShadow):
(WebCore::GraphicsContextCG::strokePath):
(WebCore::GraphicsContextCG::fillPathSegment):
(WebCore::GraphicsContextCG::strokePathSegment):
* Source/WebCore/platform/graphics/cg/GraphicsContextCG.h:
* Source/WebCore/platform/graphics/cg/PathCG.cpp:
(WebCore::PathCG::createPlatformPath):
* Source/WebCore/platform/graphics/cg/PathCG.h:
* Source/WebCore/platform/graphics/displaylists/DisplayListItems.cpp:
(WebCore::DisplayList::StrokeLine::apply const):
* Source/WebCore/platform/graphics/displaylists/DisplayListRecorder.cpp:
(WebCore::DisplayList::Recorder::fillPath):
(WebCore::DisplayList::Recorder::fillPathSegment):
(WebCore::DisplayList::Recorder::strokePath):
(WebCore::DisplayList::Recorder::strokePathSegment):
* Source/WebCore/platform/graphics/displaylists/DisplayListRecorder.h:
* Source/WebCore/platform/graphics/displaylists/DisplayListRecorderImpl.cpp:
(WebCore::DisplayList::RecorderImpl::recordStrokeLineWithColorAndThickness):
(WebCore::DisplayList::RecorderImpl::recordStrokeArcWithColorAndThickness):
(WebCore::DisplayList::RecorderImpl::recordStrokeQuadCurveWithColorAndThickness):
(WebCore::DisplayList::RecorderImpl::recordStrokeBezierCurveWithColorAndThickness):
(WebCore::DisplayList::RecorderImpl::recordStrokePathSegmentWithColorAndThickness):
(WebCore::DisplayList::RecorderImpl::recordStrokePathWithColorAndThickness):
* Source/WebCore/platform/graphics/displaylists/DisplayListRecorderImpl.h:
* Source/WebCore/rendering/svg/RenderSVGPath.cpp:
(WebCore::RenderSVGPath::updateShapeFromElement):
* Source/WebCore/rendering/svg/legacy/LegacyRenderSVGPath.cpp:
(WebCore::LegacyRenderSVGPath::updateShapeFromElement):
* Source/WebKit/GPUProcess/graphics/RemoteDisplayListRecorder.cpp:
(WebKit::RemoteDisplayListRecorder::strokeLineWithColorAndThickness):
(WebKit::RemoteDisplayListRecorder::strokeArcWithColorAndThickness):
(WebKit::RemoteDisplayListRecorder::strokeQuadCurveWithColorAndThickness):
(WebKit::RemoteDisplayListRecorder::strokeBezierCurveWithColorAndThickness):
(WebKit::RemoteDisplayListRecorder::strokePathSegmentWithColorAndThickness):
(WebKit::RemoteDisplayListRecorder::strokePathWithColorAndThickness):
* Source/WebKit/GPUProcess/graphics/RemoteDisplayListRecorder.h:
* Source/WebKit/GPUProcess/graphics/RemoteDisplayListRecorder.messages.in:
* Source/WebKit/WebProcess/GPU/graphics/RemoteDisplayListRecorderProxy.cpp:
(WebKit::RemoteDisplayListRecorderProxy::recordStrokeLineWithColorAndThickness):
(WebKit::RemoteDisplayListRecorderProxy::recordStrokeArcWithColorAndThickness):
(WebKit::RemoteDisplayListRecorderProxy::recordStrokeQuadCurveWithColorAndThickness):
(WebKit::RemoteDisplayListRecorderProxy::recordStrokeBezierCurveWithColorAndThickness):
(WebKit::RemoteDisplayListRecorderProxy::recordStrokePathSegmentWithColorAndThickness):
(WebKit::RemoteDisplayListRecorderProxy::recordStrokePathWithColorAndThickness):
* Source/WebKit/WebProcess/GPU/graphics/RemoteDisplayListRecorderProxy.h:

Canonical link: https://commits.webkit.org/269504@main
